Cleanup of docker reference manual (#625)
foundriesio/docs: source/user-guide/containers-and-docker/containers.rst

.. _ref-containers:

Containers
==========

A Factory's ``containers.git`` repo holds the source for Docker containers, as well as :doc:`compose-apps`.
When changes are made to this repo, new Targets will be built.
The CI logic for building a Target is based on simple naming rules:

 * Ignores top-level directories that ends with ``.disabled``.
 * A container image will be built for every top-level directory containing a file named ``Dockerfile``.
 * A compose app will be built for every top-level directory containing a file named ``docker-compose.yml``.

> Both a container image and compose app will be built if both a Dockerfile and Docker-compose file are included.

Reviewer: I assume both "container image" and "compose app" are terms with a notorious definition.
I could find several "container image" definitions over the ethernet, but for "compose app" is was difficult.

Reviewer 2: We are calling "Compose App" when we use a DockerCompose: https://ci.foundries.io/projects/fio-docs/builds/2478/docs/artifacts/html/user-guide/containers-and-docker/compose-apps.html

Maybe we could create a link here "Compose Apps" -> for https://ci.foundries.io/projects/fio-docs/builds/2478/docs/artifacts/html/user-guide/containers-and-docker/compose-apps.html

Advanced Container Usage
------------------------

A ``docker-build.conf`` enables advanced functionality.
This is sourced by the CI shell script and can set build time options:

 * **TEST_CMD**: Adding ``TEST_CMD="<some command>"`` directs the CI builder to run the command inside the container to verify that it is functioning correctly.

 * **SKIP_ARCHS**: A container wont be built for the specified architecture, e.g., ``SKIP_ARCHS=arm64``.

 * **EXTRA_TAGS_$ARCH**: This can work with ``SKIP_ARCHS``.

>    If builds are skipped for arm64, the arm container could be tagged for it with ``EXTRA_TAGS_arm=arm64``.

Reviewer: I don't understand this sentence.

Reviewer 2: I did not get this one either, if we skip why the arm container will be tagged? When it is tagged?

Author: I do not understand it either, I will see if I can figure out what it is getting at.

Author: I'm also putting the examples back in

Reviewer 3: I believe that means we can "reuse" `arm` builds for `arm64` - that avoids rebuilding the container for a new arch

Reviewer 4: This was a really important trick a few years ago. Nobody was producing arm64 container images. So, if you wanted to use something like the alexa SDK (32 bit only), this trick gave you a way that was previously an insanely painful task. Basically - don't build the container for arm64 (it can't be built there). But take the 32bit arm container and also tag it as arm64 (some arm platforms support this - not all) so that it can run on your 64bit OS

This produces a build command that includes ``--build-arg KEY=Value --build-arg KEY2="Value with spaces"``.

The need for dynamic arguments means some values must be generated at build time.
This can be accomplished is by taking advantage of ``docker-build.conf``;
as it is sourced by the build script, it can be used to generate content dynamically.

Example
~~~~~~~

A common case is including Git commit information into a container::

  <container>/docker-build.conf
  # $TAG is set by the build script as the Git short hash of the
  # containers.git commit being built.
  #
  # $x is the path to the container.
  cat <<EOF >$x/.docker_build_args
  GIT_SHA=$TAG
  GIT_MSG="$(git log --format=%s -1)"
  EOF

::

  <container>/Dockerfile
  ...
  # NOTE - These ARG's change *every* build. In order to maximize
  # Docker build caching, they should be as close to the end of the
  # file as possible so that the steps after these lines don't have
  # to get re-run *every* build.
  ARG GIT_SHA
  ARG GIT_MSG
  ENV GIT_SHA=$GIT_SHA
  ENV GIT_MSG=$GIT_MSG

.. _ARG:
   https://docs.docker.com/engine/reference/builder/#arg

Advanced Container Dependencies
-------------------------------

> On a rare occasion,  a Factory will need custom code to run *before* the Docker build logic is called.

Reviewer: ```suggestion
On rare occasions, a Factory will need custom code to run *before* the Docker build logic is called.
```

Reviewer 2: What about we make it more directly:
> PS. I am not sure if that is rare or not.
I understand that we want to say here is:
If you need a customer code to be executed before the docker build then you will need to ....

> This can be done with a file in the top-level directory of ``containers.git``, ``pre-build.conf``.

Reviewer: ```suggestion
This can be done by adding the file ``pre-build.conf`` in the top-level directory of ``containers.git``.
```
I almost didn't get the file name when it was at the end of the sentence.

Reviewer 2: Could we try to simplify like: `Add the file pre-build.conf in the top-level directory of containers.git:`
